File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    
    /*  A->A'->B->B'->C->C'->D->D' ,  A'->B'->C'->D' */
    
    Node* copyRandomList(Node* head) {
        
        Node *curr = head;
        
        while(curr){
           Node *list = curr->next;
           curr->next = new Node(curr->val);
           curr->next->next = list;
           curr = curr->next->next;
           curr = list;
        }
        
        curr = head;
        
        while(curr){
          Node *list = curr->next->next;
          if(curr->random){
            Node *curam = curr->random;
            curr->next->random = curam->next;
          }
         curr = list;
        }
        
        Node* ans = new Node(0);
        Node*helper = ans;
        
        while(head){
            //To make ans
            helper->next = head->next;
            helper = helper->next;
            
            //To restore head
            head->next  = head->next->next;
            head = head->next;
        }
        
       
        
        return ans->next;
           
    }
};

File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
from collections import defaultdict
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if(head == None):
            return head
        Map = defaultdict()
        curr = head
        while(curr):
            node = Node(curr.val)
            Map[curr] = node
            curr = curr.next
        
        curr = head
        
        while(curr):
            node = Map[curr]
            if(curr.next != None):
                node.next = Map[curr.next]
            if(curr.random != None):
                node.random = Map[curr.random]
            curr = curr.next
            
        curr = head
        
        return Map[curr]
        
        